whisper.cpp

```diff
@@ -6610,6 +6689,417 @@ static void whisper_exp_compute_token_level_timestamps(
     //}
 }
 
+//
+// token level timestamps - dtw version
+//
+
+// dtw + backtrace to return found path
+// based on
+// https://github.com/openai/whisper/blob/main/whisper/timing.py#L83
+static ggml_tensor * dtw_and_backtrace(ggml_context * ctx, ggml_tensor * x) {
+    WHISPER_ASSERT(x->n_dims == 2);
+
+    int64_t N = x->ne[0];
+    int64_t M = x->ne[1];
+    struct ggml_tensor * cost = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, N + 1, M + 1);
+    struct ggml_tensor * trace = ggml_new_tensor_2d(ctx, GGML_TYPE_I32, N + 1, M + 1);
+
+    cost = ggml_set_f32(cost, INFINITY);
+    trace = ggml_set_f32(trace, -1);
+    ggml_set_f32_nd(cost, 0, 0, 0, 0, 0.0);
+
+    // dtw
+    // supposedly can be optmized by computing diagonals in parallel ?
+    // Not sure it is worth it since x will be GENERATED_TOKENS*1500 size at most.
```

Comment:
How long does this step take?

Reply:
So, i've just made some measurements and it is quite rather insignificant compared to the rest of the function. 

Reporting from a Apple M2 and Metal enabled.

Just giving a very rough estimate, the total time in `whisper_exp_compute_token_level_timestamps_dtw`  seems to be dominated by the call to `whisper_decode_with_state`, followed by the call to `median_filter` and much less time on the rest.

On `medium` model, in average total function time for each 30 seconds of transcription is ~110mS. About 80mS goes to `whisper_decode_with_state`, 20mS to `median_filter` and 1mS to `dtw_and_backtrace`. 

On `large-v2` model, in average total function time for each 30 seconds of transcription is  ~230mS. About 130mS goes to `whisper_decode_with_state`, 85mS to `median_filter`, 1mS on `dtw_and_backtrace`.

Time in `dtw_and_backtrace` is practically governed by the number of tokens produced in a inference run and is independent of model size, so even in a extreme case where a run produced a lot of tokens, i don't think it would scale much compared to other times.

`whisper_decode_with_state` scales with model size. 

Most other operations in `compute_token_level_timestamps_dtw`  scale by a factor of N_TOKENS*N_ALIGNMENT_HEADS. Regarding alignment heads, the preset with most defined is the preset for large_v2, to it can be a good worst case benchmark. Nevertheless, a user setting custom alignment may use much more and it would surely heavily influence execution time.